src/runtime_src/core/pcie/tools/cloud-daemon/mpd.cpp

```diff
@@ -86,73 +119,475 @@ static int connectMsd(pcieFunc& dev, std::string ip, uint16_t port, int id)
+    }
+    case MAILBOX_REQ_USER_PROBE: {//useful for aws plugin
+        struct mailbox_conn_resp resp = {0};
+        size_t resp_len = sizeof(struct mailbox_conn_resp);
+        resp.conn_flags |= MB_PEER_READY;
+        processed = std::make_shared<sw_msg>(&resp, resp_len, orig->id(),
+            MB_REQ_FLAG_RESPONSE);
+        dev.log(LOG_INFO, "mpd daemon: response %d sent", req->req);
+        return FOR_LOCAL;
+    }
+    case MAILBOX_REQ_LOCK_BITSTREAM: {//optional
+        if (!plugin_cbs.lock_bitstream) {
+            ret = -ENOTSUP;
+            break;
+        }
+        ret = (*plugin_cbs.lock_bitstream)(dev.getIndex());
+        break;
+    }
+    case MAILBOX_REQ_UNLOCK_BITSTREAM: { //optional
+        if (!plugin_cbs.unlock_bitstream) {
+            ret = -ENOTSUP;
+            break;
+        }
+        ret = (*plugin_cbs.unlock_bitstream)(dev.getIndex());
+        break;
+    }
+    case MAILBOX_REQ_HOT_RESET: {//optional
+        if (!plugin_cbs.hot_reset) {
+            ret = -ENOTSUP;
+            break;
+        }
+        ret = (*plugin_cbs.hot_reset)(dev.getIndex());
+        break;
+    }
+    case MAILBOX_REQ_RECLOCK: {//optional
+        struct xclmgmt_ioc_freqscaling *obj =
+            reinterpret_cast<struct xclmgmt_ioc_freqscaling *>(req->data);
+        if (!plugin_cbs.hot_reset) {
+            ret = -ENOTSUP;
+            break;
+        }
+        ret = (*plugin_cbs.reclock2)(dev.getIndex(), obj);
+        break;
+    }
+    default:
+        break;
+    }
+out:
+    processed = std::make_shared<sw_msg>(&ret, sizeof(ret), orig->id(),
+        MB_REQ_FLAG_RESPONSE);
+    dev.log(LOG_INFO, "mpd daemon: response %d sent ret = %d", req->req, ret);
+    return FOR_LOCAL;
+}
+
+/*
+ * Function to notify software mailbox online/offline.
+ * This is usefull for aws. Since there is no mgmt, when xocl driver is loaded,
+ * and before the mpd daemon is running, sending MAILBOX_REQ_USER_PROBE msg
+ * will timeout and get no response, so there is no chance to know the card is
+ * ready. 
+ * With this notification, when the mpd open/close the mailbox instance, a fake
+ * MAILBOX_REQ_MGMT_STATE msg is sent to mailbox in xocl, pretending a mgmt is
+ * ready, then xocl will send a MAILBOX_REQ_USER_PROBE again. This time, mpd
+ * will get and msg and send back a MB_PEER_READY response.
+ *
+ * For other cloud vendors, as long as they want to handle mailbox msg themselves,
+ * eg. load xclbin, mpd and plugin is required. mpd exiting will send a offline
+ * notification to xocl, which will mark the card as not ready.
+ */
+int mb_notify(pcieFunc &dev, int &fd, bool online)
+{
+    struct queue_msg msg;
+    std::shared_ptr<sw_msg> swmsg;
+    std::shared_ptr<std::vector<char>> buf;
+    struct mailbox_req *mb_req = NULL;
+    struct mailbox_peer_state mb_conn = { 0 };
+    size_t data_len = sizeof(struct mailbox_peer_state) + sizeof(struct mailbox_req);
+   
+    buf = std::make_unique<std::vector<char>>(data_len, 0);
```

Comment:
will change to vector on stack

Reply:
Technically when you use a vector on the stack as you say, since this is a container, only the control part is on the stack. All the data are actually on the heap. So it does pretty much what you wanted at the first place. :-)